fem/datacollection.hpp

```diff
@@ -32,12 +32,23 @@ class DataCollection
    typedef std::map<std::string, GridFunction*> FieldMapType;
    typedef std::map<std::string, QuadratureFunction*> QFieldMapType;
 
+   /// Format constants to be used with SetFormat().
+   enum Format
+   {
+      SERIAL_FORMAT = 0, /**<
+         MFEM's serial ascii format, using the methods Mesh::Print() /
+         ParMesh::Print(), and GridFunction::Save() / ParGridFunction::Save().*/
+      PARALLEL_FORMAT = 1  /**<
+         MFEM's parallel ascii format, using the methods ParMesh::ParPrint() and
```

Comment:
I wonder if these comments are too specific.

``DataCollection`` is a base class, and derived classes might interpret these enum values differently.  

(On the other hand, ``DataCollection``, rather than ``VisItDataCollection`` is responsible for saving/loading the mesh/gridfunction files, so I can see it both ways).

Reply:
I suppose the larger question is about whether ``DataCollection`` would benefit from a refactoring that separates the stored data from how it is saved/loaded  now that there are several output formats for a ``DataCollection`` -- ``VisItDataCollection``, ``ConduitDataCollection``, ``SidreDataCollection`` -- 

Specifically, does it make sense to move the concrete Save/Load implementations for the ascii format into the derived classes. (Perhaps directly into ``VisItDataCollection``, perhaps into an ``AsciiDataCollection`` from which ``VisItDataCollection`` would derive?)

Reply:
The idea of the virtual `SetFormat()` is that:

1. the base class defines some basic formats which are given by the `Format` enum;
2. derived classes can use the formats defined in the base class, or
3. derived classes can define their own enums to use with `SetFormat()` and override `SetFormat()` and the `Save*()` methods, or
4. derived classes can ignore the values set with `SetFormat()` and provide their own methods for specifying the format.

`VisItDataCollection` takes approach 2, `SidreDataCollection` takes approach 4, and `ConduitDataCollection` does something - I'm not sure what exactly.

At this point, I'm not sure what the goal of a potential refactoring will be.

At some later time, we may want to add functionality that selects input/output formats across all the different `*DataCollection` classes allocating and initializing the appropriate (sub-)class. After that users should be able to manage the data using the common interface.

Reply:
The fields `SERIAL_FORMAT` and `PARALLEL_FORMAT` **may** be used by derived classes. That is explained here:
https://github.com/mfem/mfem/blob/881638ac85c13d183bb98c53110267f53c099613/fem/datacollection.hpp#L219-L223
I can copy that to the description of `Format` as well, if you think it will make things clearer?

I don't see how we can move the implementation of `Save()` and `Load()` out of `DataCollection` which is itself a valid class to use.

Reply:
Yes, I think adding a comment to the ``Format`` enum would clarify things a bit.
I.e. that  ``Derived classes can define their own format enumerations.``

Reply:
I updated the comment for the enum `Format`.

By the way, always feel free to extend any comments in the code that are not clear.